### ISSUE TITLE
Creating a discontinuous scale based on data I trust to be complete

### ISSUE THREAD

**Issue by ColinEberhardt** [d3fc/d3fc#1141]
_From @brentd on April 23, 2018 19:20_

First, thanks so much for open sourcing your work.

I'm a bit new to d3 and dataviz in general, so hoping for some guidance.

I have daily finance data from a stock broker's API that I'm displaying. The server is already sending me dates that I trust are "complete", where weekends and holidays are already excluded, and I just want to display it as if it were continuous. In other words, I'd rather not have the client use its own idea of what an instrument's schedule is if the server has already done the work for me.

So I came up with this using `discontinuityRange`, but it makes some obvious assumptions:

```js
// Returns a discontinuous time scale where days missing from the dataset are excluded.
function discontinuousScaleFromDays(dates) {
  const oneDay = 24 * 60 * 60 * 1000

  const ranges = dates.reduce((acc, date, i) => {
    const prev = dates[i-1]
    if (prev) {
      const distance = date.getTime() - prev.getTime()
      if (distance > oneDay) {
        const nextDay = new Date(prev.getTime() + oneDay)
        acc.push([nextDay, date])
      }
    }
    return acc
  }, [])

  return fc.scaleDiscontinuous(d3.scaleTime())
    .discontinuityProvider(fc.discontinuityRange(...ranges))
}
```

This works fine. But is there an easier way? Am I missing something?

_Copied from original issue: d3fc/d3fc-discontinuous-scale#29_

**Comment by ColinEberhardt** [comment]
_From @brentd on April 24, 2018 21:0_

A note about the above scale: while it works, I found that `discontinuityRange`, with the number of ranges I had (around 56 for 365 datapoints), is not performant enough to achieve 60fps while zooming/panning in Chrome on an i7 iMac. I'm using `transform.rescaleX()` on the discontinuous scale each frame to draw the transformed coordinates on a `canvas`.

The bulk of the time is spent in the provider's `inRange` while checking if a given date falls in a discontinuity:

```js
    const inRange = (number, range) =>
        number > range[0] && number < range[1];
```

Apparently calling `[Symbol.toPrimitive]` on the date objects to coerce them before comparing is the slow part? Kind of baffling TBH.

**Edit: indeed, modifying the provider to call `getTime()` on dates before they are compared brought time spent in `inRange` from ~10.5ms to 0.1ms. Wut.**

<img width="880" alt="screen shot 2018-04-24 at 4 54 07 pm" src="https://user-images.githubusercontent.com/1419/39213671-2f86bbea-47e0-11e8-83ca-005ee296c2b1.png">

**Comment by ColinEberhardt** [comment]
Hi @brentd thanks for the detailed issue, glad to see this project is useful for you!

Regarding your code, which creates ranges based on the gaps in your data -

> This works fine. But is there an easier way? Am I missing something?

That looks completely fine to me, and is something I've done in the past. It might be worth creating a discontinuityProvider that does exactly that, and adding it to this project.

There is one other alternative - you could just render your data with an [ordinal scale](https://github.com/d3/d3-scale#ordinal-scales), in practice, this yields a very similar result, as long as your data is already ordered.

Regarding performance, this is something we haven't tested with quite so many ranges. Maybe we should add a `dateRange` discontinuity provider that is optimised to use `getTime` for all date related maths. If you are interested in contributing, that would be great?

**Comment by ColinEberhardt** [comment]
_From @brentd on April 25, 2018 16:24_

Thanks for the response, @ColinEberhardt. I am interested in contributing those date performance tweaks and will strive to do so since it sounds like we're on the same page.

Re: ordinal scale, I just discovered this morning that it fits my use case (a discrete list of dates) - however, in practice, I'm finding it hard to work with because it doesn't implement `invert` (I guess because clamping up/down from the range to an ordinal value in the domain is too use case specific?) and thus `transform.rescaleX(scale)` doesn't work.

Ordinal scales also don't implement `ticks`, but I found that I'd have to implement ticks myself on the above discontinuous scale as well, because I want nice rounded month/year boundaries on the axis even when the 1st of the month/year falls within a discontinuation. The current behavior added after #23 filters those ticks out. But even the previous clamping behavior would require custom formatting in my case, since I want `April` on the axis and not the clamped `Apr 02`, but otherwise I do want `Apr 16` for non-month boundaries, for example.

> It might be worth creating a discontinuityProvider that does exactly that, and adding it to this project.

I might like to take a stab at this as well. Were you imagining it would use the `discontinuityRange` provider under the hood as my function does above? I'm not sure without digging in if there'd be some performance benefit of implementing a different method of distance/clamping etc when given a discrete list of dates.
